### ISSUE TITLE
Requiring status checks before merging

### ISSUE THREAD

**Issue by mahrud** [Macaulay2/M2#1783]
Github allows for restricting merges to certain branches in various ways:
- allowing only admins to merge
- requiring a certain number of approvals before allowing a merge
- requiring status checks to pass before allowing a merge

This is all documented here: https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/defining-the-mergeability-of-pull-requests

We should institute similar restrictions to prevent accidentally merging PR's with failing tests into incorrect branches, as happened today with #1778.

To make matters worse, after that PR was merged into master, master was merged into release-1.17, then release-1.17 was merged into development and pre-master. This is excessive and confusing, and has lead to **all four active branches** to be in a failing state!

Moreover, now that there's a fix (#1782), even fixing it will leads to a spaghetti of merges between these 4 branches, because we have to do 4 merges to fix all the branches! I think we should also reduce the number of parallel active branches to only master and development. As I explained in https://github.com/Macaulay2/M2/issues/1691#issuecomment-747184042:

>  If we use tags instead of branches for each release, including release candidates, then I don't see a reason to keep the `release-*` branches: whenever you have a commit that's reasonably ready, tag it `release-1.17-rc1`, when new issues are found and fixed, tag `release-1.17-rc2`, until it's ready to be tagged `release-1.17`.

**Comment by DanGrayson** [comment]
Yes, I keep mistakenly merging into "master", as it is natural for users to direct pull requests there.   It would be great if I could do that only after being presented with a vivid warning, for any branch that is not "development".

**Comment by mahrud** [comment]
Sure. If we follow some basic etiquette, we can also avoid such objective messiness:
![image](https://user-images.githubusercontent.com/3147718/104851832-00c66f80-58bd-11eb-8f31-0c2874e468c7.png)

I provided the link to the documentation above.

**Comment by DanGrayson** [comment]
I've been waiting for that magic button to suddenly appear for a week now.  Thanks!

**Comment by mahrud** [comment]
You hadn't merged into master till recently, I believe.

**Comment by DanGrayson** [comment]
I think it's because I never happened to click on one of these two "workflows":

<img width="411" alt="Screen Shot 2021-01-22 at 1 34 08 PM" src="https://user-images.githubusercontent.com/700228/105536852-9f384380-5cb6-11eb-9168-0ff1affb4bf6.png">

**Comment by DanGrayson** [comment]
I've just figured out a potentially better way: click a button to submit a pull request from development to master.  As in https://github.com/Macaulay2/M2/pull/1862

**Comment by mahrud** [comment]
@DanGrayson @mikestillman you may have seen the following tip from GitHub:
![image](https://user-images.githubusercontent.com/3147718/181045905-c61d09bb-6d8b-40a4-9891-d9e3e58f06c4.png)

**Comment by mahrud** [comment]
Updating for new features: GitHub's [branch protection settings](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branches#about-branch-protection-settings) have quite a few options that would work well for us:
- require reviews,
- require status checks (e.g. green checks),
- require a conversation, etc.

In addition, GitHub [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) files tell it who "owns" a piece of code and should submit a review before it is merged. I'm not sure if we necessarily need this, but it's an option.

cc: @antonleykin @mikestillman @d-torrance
